src/app/vertextool/qgsvertextool.cpp

```diff
@@ -545,31 +548,71 @@ void QgsVertexTool::cadCanvasReleaseEvent( QgsMapMouseEvent *e )
         continue;
       }
 
+      QgsRenderContext context = QgsRenderContext::fromMapSettings( mCanvas->mapSettings() );
+      context.expressionContext() << QgsExpressionContextUtils::layerScope( vlayer );
+      std::unique_ptr< QgsFeatureRenderer > r;
+      if ( vlayer->renderer() )
+      {
+        r.reset( vlayer->renderer()->clone() );
+        r->startRender( context, vlayer->fields() );
+      }
+
       QgsRectangle layerRect = layerRubberBandGeometry.boundingBox();
       std::unique_ptr< QgsGeometryEngine > layerRubberBandEngine( QgsGeometry::createGeometryEngine( layerRubberBandGeometry.constGet() ) );
       layerRubberBandEngine->prepareGeometry();
 
+      QgsFeatureRequest request;
+      request.setFilterRect( layerRect );
+      request.setFlags( QgsFeatureRequest::ExactIntersect );
+      if ( r )
+        request.setSubsetOfAttributes( r->usedAttributes( context ), vlayer->fields() );
+      else
+        request.setNoAttributes();
+
       QgsFeature f;
-      QgsFeatureIterator fi = vlayer->getFeatures( QgsFeatureRequest( layerRect ).setNoAttributes() );
+      QgsFeatureIterator fi = vlayer->getFeatures( request );
       while ( fi.nextFeature( f ) )
       {
         if ( mLockedFeature && mLockedFeature->featureId() != f.id() )
           continue;  // with locked feature we only allow selection of its vertices
 
+        context.expressionContext().setFeature( f );
+        // make sure to only use features that are visible
+        bool isFeatureInvisible = ( r && !r->willRenderFeature( f, context ) );
+        if ( isFeatureInvisible && invisibleVertexSelected )
+          continue;
```

Comment:
It took me a while to understand this logic. Maybe a comment like:

 // If this feature is not visible, and we've already determined that we have to show users a warning about selecting
 // invisible vertices, then we don't need to check through the vertices for this feature. We'll be showing the warning anyway regardless of the outcome!

Reply:
OK! Also renamed `invisibleVertexSelected' to `showInvisibleFeatureWarning`. I think the intention is clearer now!

Reply:
Much better, thanks!